clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1330,6 +1330,27 @@ event_pre_syscall(void *drcontext, int sysnum)
         return true;
     if (BUF_PTR(data->seg_base) == NULL)
         return true; /* This thread was filtered out. */
+
+    // If we just switched modes in this thread in a block ending in a syscall,
+    // we can end up here without having traced the syscall instr, which we
+    // want to avoid.  We look for an empty (non-initial, so we do not add
+    // init_header_size) buffer with zero instrs in the current window (b/c it
+    // could be empty just due to the syscall instr filling up the prior buffer).
```

Comment:
This issue (markers etc without the corresponding instr) can happen for other cases where the markers are added by a callback that is invoked even in the no-trace mode. I wonder if we have any other such cases.

Reply:
They all have a mode check: event_pre_syscall, event_post_syscall, event_kernel_xfer.  We want the marker anyway for event_thread_exit.